taichi/program/program.cpp

```diff
@@ -73,53 +73,44 @@ Program::Program(Arch desired_arch)
 #ifdef TI_WITH_LLVM
     program_impl_ = std::make_unique<LlvmProgramImpl>(config, profiler.get());
 #else
-    TI_NOT_IMPLEMENTED
+    TI_ERROR("This taichi is not compiled with LLVM");
 #endif
   } else if (config.arch == Arch::metal) {
     if (!metal::is_metal_api_available()) {
-      TI_WARN("No Metal API detected.");
-      config.arch = host_arch();
+      TI_ERROR("No Metal API detected.");
     } else {
       program_impl_ = std::make_unique<MetalProgramImpl>(config);
     }
-  }
+  } else if (config.arch == Arch::vulkan) {
 #ifdef TI_WITH_VULKAN
-  else if (config.arch == Arch::vulkan) {
     if (!vulkan::is_vulkan_api_available()) {
-      TI_WARN("No Vulkan API detected.");
-      config.arch = host_arch();
+      TI_ERROR("No Vulkan API detected.");
     } else {
       program_impl_ = std::make_unique<VulkanProgramImpl>(config);
     }
-  }
+#else
+    TI_ERROR("This taichi is not compiled with Vulkan")
 #endif
-
-  if (config.arch == Arch::opengl) {
+  } else if (config.arch == Arch::opengl) {
     if (!opengl::is_opengl_api_available()) {
-      TI_WARN("No OpenGL API detected.");
-      config.arch = host_arch();
+      TI_ERROR("No OpenGL API detected.");
     } else {
       program_impl_ = std::make_unique<OpenglProgramImpl>(config);
     }
-  }
-
-  if (config.arch == Arch::cc) {
+  } else if (config.arch == Arch::cc) {
 #ifdef TI_WITH_CC
     program_impl_ = std::make_unique<CCProgramImpl>(config);
 #else
-    TI_WARN("No C backend detected.");
-    config.arch = host_arch();
+    TI_ERROR("No C backend detected.");
 #endif
+  } else {
+    TI_NOT_IMPLEMENTED
   }
 
-  if (config.arch != desired_arch) {
-    TI_WARN("Falling back to {}", arch_name(config.arch));
-  }
+  TI_ERROR_IF(!program_impl_, "Backend not found.");
```

Comment:
ditto :D

Reply:
Would you mind also adding a comment about invariant program_impl_ is not empty?